### DISCUSSION TITLE
typing a result where the column is optional, but the result is not

### DISCUSSION THREAD

**Discussion by jvanasco** [sqlalchemy/sqlalchemy#11390] [Usage Questions] [answered]
Assuming we have a mapped class with an optional attrib / nullable column:

    class User(DeclaredTable):
        __tablename__ = "user"

        id: Mapped[int] = mapped_column(Integer, primary_key=True)
        name: Mapped[Optional[str]] = mapped_column(Unicode, nullable=True)

Executing the following two queries against this - which imply/require a value in the returned row - are type hinted with the column/attrib being potentially null:

    s = Session()

    q1 = s.query(User.name).filter(
        sqlalchemy_func.lower(User.name).startswith("foo"),
    )
    q2 = s.query(User.name).filter(
        sqlalchemy_func.lower(User.name).startswith("foo"),
        User.name.is_not(None),
    )

    r1 = q1.all()
    reveal_type(r1)

    r2 = q2.all()
    reveal_type(r2)

both show the same:

> Revealed type is "builtins.list[sqlalchemy.engine.row.Row[tuple[Union[builtins.str, None]]]]"
> Revealed type is "builtins.list[sqlalchemy.engine.row.Row[tuple[Union[builtins.str, None]]]]"

Is there a cheap way to assert or cast this value to *not* be optional?  The only approaches I can think of involve processing the entire list.

The context for my usage:

I often have an internal API function that will return not-None values, which is backed by a sqlalchemy db query that involves nullable columns, but the search only queries populated columns.

**Comment by CaselIT** [comment, marked as answer]
Hi,

We added these. Nullable and NotNullable: https://docs.sqlalchemy.org/en/20/changelog/changelog_20.html#change-2c09b04320953f54a223ad0d14d16794

**Reply by jvanasco** [reply]
Thank you so much!  This is perfect.
